### ISSUE TITLE
engine.get_triggered_abbreviation() returns empty abbreviation when “Omit trigger character” is enabled

### ISSUE THREAD

**Issue by muekoeff** [autokey/autokey#721]
### Has this issue already been reported?

- [X] I have searched through the existing issues.

### Is this a question rather than an issue?

- [X] This is not a question.

### What type of issue is this?

Bug

### Which Linux distribution did you use?

Linux 5.15.57-2-MANJARO

### Which AutoKey GUI did you use?

GTK

### Which AutoKey version did you use?

0.96.0

### How did you install AutoKey?

AUR [autokey-gtk](https://aur.archlinux.org/packages/autokey-gtk)

### Can you briefly describe the issue?

When running a script with the source code

```python
import json

abbreviation, trigger = engine.get_triggered_abbreviation()
dialog.info_dialog('Debug', json.dumps(engine.get_triggered_abbreviation()))
```
[Test.zip](https://github.com/autokey/autokey/files/9274057/Test.zip)

by triggering its abbreviation (for which “Omit trigger character” is enabled) `engine.get_triggered_abbreviation()` returns a tuple of two empty strings.
For this script a dialog showing `["", ""]` is displayed.

### Can the issue be reproduced?

Always

### What are the steps to reproduce the issue?

1. Create a new script that uses `engine.get_triggered_abbreviation()`
2. Create an abbreviation and enable “Omit trigger character”
3. Type abbreviation

### What should have happened?

`engine.get_triggered_abbreviation()[0]` should return that entered abbreviation

### What actually happened?

`engine.get_triggered_abbreviation()` returns a tuple of two empty strings: `["", ""]`

### Can you provide the output of the AutoKey command?

Abbreviation `##foobar`
```bash
autokey.service - Raw key: '#', modifiers: [], Key: #
autokey.service - Window visible title: 'New Issue · autokey/autokey — Mozilla Firefox', Window class: 'Navigator.firefox'
autokey.service - Input queue at end of handle_keypress: deque(['#'], maxlen=150)
autokey.service - Raw key: '#', modifiers: [], Key: #
autokey.service - Window visible title: 'New Issue · autokey/autokey — Mozilla Firefox', Window class: 'Navigator.firefox'
autokey.service - Input queue at end of handle_keypress: deque(['#', '#'], maxlen=150)
autokey.service - Raw key: 'f', modifiers: [], Key: f
autokey.service - Window visible title: 'New Issue · autokey/autokey — Mozilla Firefox', Window class: 'Navigator.firefox'
autokey.service - Input queue at end of handle_keypress: deque(['#', '#', 'f'], maxlen=150)
autokey.service - Raw key: 'o', modifiers: [], Key: o
autokey.service - Window visible title: 'New Issue · autokey/autokey — Mozilla Firefox', Window class: 'Navigator.firefox'
autokey.service - Input queue at end of handle_keypress: deque(['#', '#', 'f', 'o'], maxlen=150)
autokey.service - Raw key: 'o', modifiers: [], Key: o
autokey.service - Window visible title: 'New Issue · autokey/autokey — Mozilla Firefox', Window class: 'Navigator.firefox'
autokey.service - Input queue at end of handle_keypress: deque(['#', '#', 'f', 'o', 'o'], maxlen=150)
autokey.service - Raw key: 'b', modifiers: [], Key: b
autokey.service - Window visible title: 'New Issue · autokey/autokey — Mozilla Firefox', Window class: 'Navigator.firefox'
autokey.service - Input queue at end of handle_keypress: deque(['#', '#', 'f', 'o', 'o', 'b'], maxlen=150)
autokey.service - Raw key: 'a', modifiers: [], Key: a
autokey.service - Window visible title: 'New Issue · autokey/autokey — Mozilla Firefox', Window class: 'Navigator.firefox'
autokey.service - Input queue at end of handle_keypress: deque(['#', '#', 'f', 'o', 'o', 'b', 'a'], maxlen=150)
autokey.service - Raw key: 'r', modifiers: [], Key: r
autokey.service - Window visible title: 'New Issue · autokey/autokey — Mozilla Firefox', Window class: 'Navigator.firefox'
autokey.service - Input queue at end of handle_keypress: deque(['#', '#', 'f', 'o', 'o', 'b', 'a', 'r'], maxlen=150)
autokey.service - Raw key: ' ', modifiers: [], Key:  
autokey.service - Window visible title: 'New Issue · autokey/autokey — Mozilla Firefox', Window class: 'Navigator.firefox'
autokey.service - Matched Script "Test" having abbreviations "['##foobar']" against current input
autokey.service - Script runner executing: Script('Test')
autokey.service - Triggered a Script by an abbreviation. Setting it for engine.get_triggered_abbreviation(). abbreviation='', trigger=''
autokey.interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
autokey.service - Input queue at end of handle_keypress: deque([], maxlen=150)
autokey.interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
autokey.interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
autokey.interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
autokey.interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
autokey.interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
autokey.interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
autokey.interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
autokey.interface - Send special key: [<Key.BACKSPACE: '<backspace>'>]
autokey.service - Raw key: <Key.ENTER: '<enter>'>, modifiers: [], Key: Key.ENTER
autokey.service - Window visible title: 'Debug', Window class: 'zenity.Zenity'
autokey.service - Input queue at end of handle_keypress: deque(['\n'], maxlen=150)
```

**Comment by Elliria** [comment]
I could be wrong, but I suspect this may be related to the discussion we had in [Gitter](https://gitter.im/autokey/autokey) a while back about persistent values (**local values** and **global values**) not immediately being written to the files they eventually end up in and, instead, temporarily being in some as-yet-unknown place until something further occurs (like another saved change to the program or the closing of the program). These probably also are held somewhere temporarily and are, at least for the moment, inaccessible while there. Like I said, I could be wrong, but my Spidey Senses are tingling.

**Comment by josephj11** [comment]
@muekoeff Until this bug is resolved, here's a workaround. It has some edge cases, but works in general.

Neat trick from William Morris on our list: When you trigger a script with an abbreviation, you can turn off the delete the abbreviation switch. Once you do that, you can shift left the appropriate number of times to highlight the abbreviation and then copy or cut it to your clipboard. Now you know what abbreviation triggered your script. If the script has multiple abbreviations as triggers, you can use them to modify its behavior. You can also use Ctrl+Shift+Left to select the entire word to the left to get the whole trigger abbreviation without knowing in advance how long it is.

**Comment by randolf-scholz** [comment]
The bug is likely this line:

https://github.com/autokey/autokey/blob/58d488bb0160953b9eb4b6b571dfc01b729e42c2/lib/autokey/service.py#L563

when `len(trigger_character)` is `0`, this becomes `buffer[:-0]` which is equivalent to `buffer[:0]`, yielding an empty string. 

a safe way of removing potentially empty suffix would be to use `buffer[:-len(trigger_character) or None]` or even better `buffer.removesuffix(trigger_character)`.

**Comment by josephj11** [comment]
@randolf-scholz Thanks! We'll take a look at this.

@chrislee35 This might be an easy fix.

**Comment by josephj11** [comment]
@randolf-scholz Damn! You have a whole (small) [project](https://github.com/randolf-scholz/autokey_julia_unicode) based on AutoKey and didn't tell us about it. We love to hear from users. BTW, our develop branch also works on Wayland and we hope to release it as 0.97.0 in the near future. It still needs testing and may have bugs RN.

**Comment by chrislee35** [comment]
@randolf-scholz's diagnosis was spot on for the code as it stood when that blob link was made, but it points at a stale commit (`58d488b`) — `develop` has since moved past it.

The buggy `buffer[:-len(trigger_character)]` line was already replaced back in `9ad3dbc` ("fix triggered abbreviation for immediate scripts", merged 2026-05-11), which switched to `script._partition_input(buffer, configured_abbreviation)` — finding the abbreviation by matching it in the buffer instead of subtracting the trigger character's length. That fix was aimed at a different setting (`immediate`), but it happens to share the same code path, so it fixed this too as a side effect.

I confirmed against current `develop` with the actual reported setting (`omitTrigger=True`, abbreviation `##foobar`), not just the `immediate` case the existing test covers:

```
engine._set_triggered_abbreviation called with: call('##foobar', '')
```

Correct abbreviation comes through, not `''`. This is fixed and safe to close.

Looks like we found a power-user and a premium use case for AutoKey. This may want to consider building sharable libraries of various replacements for different domains, not just math.

**Comment by josephj11** [comment]
Closing as already fixed.

@randolf-scholz Thanks for the submission!

**Comment by randolf-scholz** [comment]
@chrislee35 ah ok, I found this on my local version (0.96), and [code search on github](https://github.com/search?q=repo%3Aautokey%2Fautokey+%22triggered_abbreviation+%3D+buffer%5B%3A-len%28trigger_character%29%5D%22&type=code) showed the exact same line present on the master branch.

@josephj11 We actually already had a brief exchange a few years ago (#599). If you have like an official list of projects I'd glady add my repo to that list. I can also give some feedback, is https://github.com/autokey/autokey/discussions/categories/beta-feedback the right place for that?

Excited for wayland support. The lack of working text expander (the last time I tried a few years back) still kept me back using X11. 

Though it seems you now have tough competition with [espanso](https://github.com/espanso/espanso), and I have been seriously considering moving my own setup there (due to claimed wayland support). Though I think in my project both can live side-by-side  since they would only require different install script, albeit [there already seem to exist similar projects for espanso](https://hub.espanso.org/search/?t=math).

**Comment by josephj11** [comment]
@randolf-scholz A list of featured projects would be a nice addition to our wiki. I think it would fit nicely in the Community section. (Anyone with a GitHub account can edit the wiki)... (hint :)  )

Feedback depends on what type it is.
- So far, the Beta discussion has only been used when we have a Beta (hopefully, very soon)
- For little things that don't need to be on view as part of the project, [Gitter](https://gitter.im/autokey/autokey) is best (I think) more users look there
- The General Discussion category would be fine too
- We can start a new category if you have a good idea for one

I haven't looked at Espanso in a while, but last I looked, they were very good with phrases, but didn't have much in the way of scripting.

**Comment by chrislee35** [comment]
I'll look into Espanso as well, just so I know when it meets a user's need better than AutoKey.

**Comment by josephj11** [comment]
@chrislee35 See this [discussion](https://github.com/autokey/autokey/discussions/775) for other AutoKey alternatives.

**Comment by chrislee35** [comment]
I like the style and detail level of the wiki.  Thanks for the link.

**Comment by josephj11** [comment]
Oops. That's a discussion thread. I started it because we get questions about alternatives and to see what our competitors are up to.

We saw that at least phrases would work on Wayland because Espanso had that before we started.

As for the real wiki, @Elliria and I and a few others put a lot of work into the wiki. It is really useful. It's also great to be able to answer someone with a link instead of explaining something over and over.

Maybe it would also be good training data for an AI assistant.
